open-craft/configuration
# [BB-2217] Allow setting SiteConfiguration values when provisioning an instance

This PR implements the changes to allow setting `SiteConfiguration`values in LMS and Studio when provisioning an instance.

**Testing instructions**:
1. Provision an instance with this branch. Verify that there are no errors.
1. Provision a new appserver on the instance with `EDXAPP_SITE_CONFIGURATION` set (see the comment in the `playbooks/roles/edxapp/defaults/main.yml` file for the structure of values). Verify that the `SiteConfiguration` values are set and there are no errors.
  For example, this could be a good test configuration
    ```yaml
   EDXAPP_SITE_CONFIGURATION:
     - values:
        ENABLE_ACCOUNT_DELETION: false
        ENABLE_SOCIAL_MEDIA_LINKS: true
    - values:
        ENABLE_ACCOUNT_DELETION: true
        ENABLE_SOCIAL_MEDIA_LINKS: false
      site_id: 1
    - values:
        ENABLE_ACCOUNT_DELETION: false
        ENABLE_SOCIAL_MEDIA_LINKS: true
      domain: example.com
    ```
1. Repeat the previous step by adding another item to the list with both `site_id` and `domain` keys set and verify that the playbook fails with an error message stating that both cannot be specified at the same time.

Since each of the above steps involve provisioning new appservers and that can take a lot of time, there is a workaround to test the steps 2-4.
1. Log in to the appserver created for the step 1.
1. Navigate to the `/edx/app/edx_ansible/edx_ansible` directory as the `root` user. Create a Python 2.7 virtualenv environment, activate it and install the requirements in `requirements.txt`.
1. From the `Combined ansible vars` section of the appserver successfully provisioned in the first step, copy all the generated ansible variables.
1. Save the variables on disk (say `/tmp/vars.yml`).
1.  Edit the `/edx/app/edx_ansible/edx_ansible/playbooks/openedx_native.yml` playbook to comment out the entire `Bootstrap instance(s)` play.
1. Under the `Configure instance(s)` play, add `connection: local` before the `hosts: all` line.
1. The edits done in the previous steps will reduce the running time of the playbook by commenting out the tasks/plays that were already run successfully and hence need not be run repeatedly during testing.
1. Run the following command from the `/edx/app/edx_ansible/edx_ansible/playbooks` directory, replacing the name and path to the vars file with the correct ones you have used.

    ```
    ansible-playbook -i "127.0.0.1," -e @/tmp/vars.yml openedx_native.yml --tags create_or_update_configuration
    ```
1. For testing various scenarios, update the vars file, update/delete the existing `SiteConfiguration` values as required and re-run the command mentioned in the previous step to test various scenarios.

Configuration Pull Request
---

Make sure that the following steps are done before merging
